### PR TITLE
thing-shadow

### PR DESCRIPTION
**Pull request by liaojason2softchef**

### CODE REVIEW

File: lambda-assets/things/delete-thing-shadow/app.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { IoTDataPlaneClient, DeleteThingShadowCommand } from "@aws-sdk/client-iot-data-plane";
```
Comment (MinCheTsai) [inline lambda-assets/things/delete-thing-shadow/app.ts:1]: lint

File: lambda-assets/things/update-thing-shadow/app.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { IoTDataPlaneClient, UpdateThingShadowCommand } from "@aws-sdk/client-iot-data-plane";
```
Comment (MinCheTsai) [inline lambda-assets/things/update-thing-shadow/app.ts:1]: lint

File: lambda-assets/things/list-thing-shadows/app.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { IoTDataPlaneClient, ListNamedShadowsForThingCommand } from "@aws-sdk/client-iot-data-plane";
```
Comment (MinCheTsai) [inline lambda-assets/things/list-thing-shadows/app.ts:1]: lint

File: lambda-assets/things/get-thing-shadow/app.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { IoTDataPlaneClient, GetThingShadowCommand } from "@aws-sdk/client-iot-data-plane";
```
Comment (MinCheTsai) [inline lambda-assets/things/get-thing-shadow/app.ts:1]: lint

File: lambda-assets/things/get-thing-shadow/app.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        thingName: request.parameter('thingName'),
+        shadowName: request.parameter('shadowName'),
+      }),
+    );
+    let payloadString: string = '';
+    payload.forEach(num => {
+      payloadString += String.fromCharCode(num);
+    });
+    return response.json({
+      payloadString
```
Comment (MinCheTsai) [inline lambda-assets/things/get-thing-shadow/app.ts:20]: 命名需要調整

File: lambda-assets/things/list-thing-shadows/app.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    };
+    const ioTDataPlaneClient = new IoTDataPlaneClient({});
+    const thingShadows = await ioTDataPlaneClient.send(
+      new ListNamedShadowsForThingCommand({
+        thingName: request.parameter('thingName'),
+        ...parameters,
+      })
+    )
+    return response.json({
+      thingShadows
```
Comment (MinCheTsai) [inline lambda-assets/things/list-thing-shadows/app.ts:20]: 資料結構需要確認

File: lambda-assets/things/update-thing-shadow/app.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import { Request, Response } from '@softchef/lambda-events';
+
+export async function handler(event: { [key: string]: any }) {
+  const request = new Request(event);
+  const response = new Response();
+  try {
+    const validated = request.validate(joi => {
+      return {
+        payload: joi.object().required(),
```
Comment (MinCheTsai) [inline lambda-assets/things/update-thing-shadow/app.ts:10]: shadow payload rule?

File: lambda-assets/things/update-thing-shadow/app.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+export async function handler(event: { [key: string]: any }) {
+  const request = new Request(event);
+  const response = new Response();
+  try {
+    const validated = request.validate(joi => {
+      return {
+        payload: joi.object().keys({
+          state: joi.object().keys({
+            desired: joi.object().required(),
```
Comment (MinCheTsai) [inline lambda-assets/things/update-thing-shadow/app.ts:12]: allow null

Comment (liaojason2softchef) [reply lambda-assets/things/update-thing-shadow/app.ts:12]: payload is required in UpdateThingShadowCommand.

Comment (MinCheTsai) [reply lambda-assets/things/update-thing-shadow/app.ts:12]: desired & reported are optional

**Review by MinCheTsai — APPROVED** [review]

**Pull request merged**